### PR TITLE
Adds Local Zlevel Drift

### LINKED ISSUE #4319: Obfuscating map coordinates
Currently all coordinates on each z level are always the same, this leads to telescience spreadsheets or coordinate hunting loot with GPS and a jetpack.

The player can see coordinates ingame using various tools like the GPS or crew monitor.

What I propose is that there would be a random offset generated each round for each z level that only visually affects the coordinates player sees / inputs, that is internally as far as game is concerned all coordinates would remain as they are now.

However displaying coordinates or inputing them ingame from the players perspective would always be with the added offset. Same concept could be applied to z level numbers.

This would mean that at the very least a player would have to first explore a z level to know where he is, find a point of reference for his spreadsheet and figure out the rest.

### PR DESCRIPTION
**Pull request by clusterfack**
Resolves #4319.

Adds a zlevel drift for all zlevels at roundstart. Hopefully it doesn't go list index out of bounds or something. Also stopped a method for determining zlevel drift using only the teleport computer and inputting 6 values to bisect the possible offsets until you find the specific one.

### CODE REVIEW

**Comment by Exxion** [comment]
Wasn't this already a thing? I'm about 95% sure this was already implemented, albeit with the same random offset applied to all Z-levels.

**Comment by PJB3005** [comment]
Yeah, I think this was a thing too, I've always had issues with Tsciencing to the wrong location, even though my coords in the GPS are correct

**Comment by TermedClepe** [comment]
Wouldn't it be more manageable to create a function for getting real / obfuscated coordinates instead of calculating it all over the place each time?

File: code/game/machinery/doppler_array.dm [hunk 1]
```diff
@@ -125,9 +125,9 @@ var/list/doppler_arrays = list()
 	if(!(direct & dir))	return
 	*/
 
-	var/message = "Explosive disturbance detected - Epicenter at: grid ([x0-WORLD_X_OFFSET],[y0-WORLD_Y_OFFSET], [z0]). [cap ? "\[Theoretical Results\] " : ""]Epicenter radius: [devastation_range]. Outer radius: [heavy_impact_range]. Shockwave radius: [light_impact_range]. Temporal displacement of tachyons: [took] second\s.  Data logged."
+	var/message = "Explosive disturbance detected - Epicenter at: grid ([x0-WORLD_X_OFFSET[z0]],[y0-WORLD_Y_OFFSET[z0]], [z0]). [cap ? "\[Theoretical Results\] " : ""]Epicenter radius: [devastation_range]. Outer radius: [heavy_impact_range]. Shockwave radius: [light_impact_range]. Temporal displacement of tachyons: [took] second\s.  Data logged."
```
Comment (ComicIronic) [inline code/game/machinery/doppler_array.dm]: Is z0 defined anywhere?

Comment (PJB3005) [inline code/game/machinery/doppler_array.dm]: well it's compiling, so yes?

Comment (ComicIronic) [inline code/game/machinery/doppler_array.dm]: Byond has a funny way of dealing with lists, which is that any index which would be valid as a var name i.e. "z0" can be put bare into the list.

`list(z0 = whatever)`
would compile, for example.

**Comment by DeityLink** [comment]
And all the telescientists cross the land cried out in unisson.

File: code/game/machinery/doppler_array.dm (hunk 1 above)
Comment (clusterfack) [inline code/game/machinery/doppler_array.dm]: z0 is literally used just to the right of where I placed this.

**Comment by clusterfack** [comment]
LOCAL drift. Each zlevel has a different offset. Since the first zlevel's offset could be figured out instantaneously simply by walking to a standard location and measuring, it was previously ridiculously easy to subvert.

**Comment by clusterfack** [comment]
If you teleport a gps to 150,150,4. It will end up on 150,150,4.
If you see a dead body on a crew monitoring computer at 150,150,4. It will be at 150,150,4.

But if you use the main station to find the offset and it's offset is 34,-23. If you adjusted for that you still wouldn't end up at the same location every time like you do currently, because the derelict has a unique drift that would put you somewhere completely different.

**Comment by clusterfack** [comment]
Currently runtiming on world/new, I'll fix it

**Comment by Shadowmech88** [comment]
So essentially you just have to calibrate it for each Z-level now instead of calibrating it once for all Z-levels?

**Comment by clusterfack** [comment]
Well there is calibration, which is just error in the machine from -10 to 10. But then there is the 'true' location of objects which depended on just the gps coordinates minus the world drift, but now depends on the drift of the zlevel you are aiming for. 

So if you want to teleport to your spreadsheet of 'true' locations, you need to find the drift for that zlevel.

**Comment by DeityLink** [comment]
Just to be sure, local zDrift is world drift right? So GPS and suit sensors won't be wrong right?

**Comment by clusterfack** [comment]
A GPS and suit sensors would give you the correct coordinates. The only case in which telescientists are cockblocked is if they are trying to find the "world drift" so they can beam up random objects from every other zlevel without a frame of reference on that level.

**Comment by DeityLink** [comment]
Alright, then I'm fine with it.

**Comment by clusterfack** [comment]
This is finished

**Pull request merged**